File: backend/app/schemas/analytics.py
import uuid
from datetime import date
from decimal import Decimal

from pydantic import BaseModel

from app.models.enums import RecurringFrequency, TransactionType
from app.schemas.category import CategoryRead


class SummaryRead(BaseModel):
    month: int
    year: int
    income: Decimal
    expenses: Decimal
    net_cash_flow: Decimal
    previous_month_income: Decimal
    previous_month_expenses: Decimal
    income_change_percent: float | None
    expenses_change_percent: float | None


class CategoryBreakdownItem(BaseModel):
    category: CategoryRead
    amount: Decimal
    transaction_count: int
    percent_of_total: float


class TrendItem(BaseModel):
    year: int
    month: int
    income: Decimal
    expenses: Decimal
    net_cash_flow: Decimal


class MerchantItem(BaseModel):
    payee: str
    total_amount: Decimal
    transaction_count: int


class RecurringSeriesItem(BaseModel):
    payee: str
    category: CategoryRead
    type: TransactionType
    amount: Decimal
    frequency: RecurringFrequency
    last_date: date
    next_due_date: date
    transaction_id: uuid.UUID


class RecurringAnalysisRead(BaseModel):
    series: list[RecurringSeriesItem]
    total_monthly_estimate: Decimal

File: frontend/src/types/analytics.ts
import type { Category, CategoryType } from './category'
import type { RecurringFrequency } from './transaction'

export interface Summary {
  month: number
  year: number
  income: string
  expenses: string
  net_cash_flow: string
  previous_month_income: string
  previous_month_expenses: string
  income_change_percent: number | null
  expenses_change_percent: number | null
}

export interface CategoryBreakdownItem {
  category: Category
  amount: string
  transaction_count: number
  percent_of_total: number
}

export interface TrendItem {
  year: number
  month: number
  income: string
  expenses: string
  net_cash_flow: string
}

export interface MerchantItem {
  payee: string
  total_amount: string
  transaction_count: number
}

export interface RecurringSeriesItem {
  payee: string
  category: Category
  type: CategoryType
  amount: string
  frequency: RecurringFrequency
  last_date: string
  next_due_date: string
  transaction_id: string
}

export interface RecurringAnalysis {
  series: RecurringSeriesItem[]
  total_monthly_estimate: string
}
